stack/eslint-patch/src/modern-module-resolution.ts

```diff
@@ -31,17 +31,34 @@ for (let currentModule = module; ; ) {
   if (!eslintrcBundlePath) {
     // For ESLint >=8.0.0, all @eslint/eslintrc code is bundled at this path:
     //   .../@eslint/eslintrc/dist/eslintrc.cjs
-    if (/[\\/]@eslint[\\/]eslintrc[\\/]dist[\\/]eslintrc\.cjs$/i.test(currentModule.filename)) {
-      const eslintrcFolder: string = path.join(path.dirname(currentModule.filename), '..');
-      eslintrcBundlePath = path.join(eslintrcFolder, 'dist/eslintrc.cjs');
-    }
+    try {
+      const eslintrcFolder = path.dirname(
+        require.resolve('@eslint/eslintrc/package.json', { paths: [currentModule.path] })
+      );
+
+      // Make sure we actually resolved the module in our call path
+      // and not some other spurious dependency.
+      if (path.join(eslintrcFolder, 'dist/eslintrc.cjs') == currentModule.filename) {
+        eslintrcBundlePath = path.join(eslintrcFolder, 'dist/eslintrc.cjs');
+      }
+    } catch {}
   } else {
     // Next look for a file in ESLint's folder
     //   .../eslint/lib/cli-engine/cli-engine.js
-    if (/[\\/]eslint[\\/]lib[\\/]cli-engine[\\/]cli-engine\.js$/i.test(currentModule.filename)) {
-      eslintFolder = path.join(path.dirname(currentModule.filename), '../..');
-      break;
-    }
+    try {
+      const eslintCandidateFolder = path.dirname(
+        require.resolve('eslint/package.json', {
+          paths: [currentModule.path]
+        })
+      );
+
+      // Make sure we actually resolved the module in our call path
+      // and not some other spurious dependency.
+      if (path.join(eslintCandidateFolder, 'lib/cli-engine/cli-engine.js') == currentModule.filename) {
```

Comment:
(Generally we use `===` instead of `==`.)

However this comparison seems like it might fail on Windows, where filenames are case-insensitive and path separators are sometimes `\` instead of `/`

Reply:
Oops, I'm getting too habituated to a linter yelling at me for that kind of thing.

Re path separators, those are normalized by path.join:
```
> path.join("foo", "bar/baz")
'foo\\bar\\baz'
```
Case sensitivity is interesting though. Do we have reason to suspect that eslint installations will randomly have upper case letters in file names. We can obviously normalize casing before comparison, but that feels a bit unnecessary.

Reply:
I suppose you have a point that these files get extracted from ESLint's NPM package, so their filenames should have hard-wired case.  (The way we usually encounter this problem is when a developer renames their source file to have different case, but imports or caches are still using the original casing.)  So yeah, in this particular case it probably isn't a concern.

Reply:
@asmundg your latest iteration LGTM. Is there anything remaining before we merge it?

Reply:
@octogonz no, I'm pretty happy now. :)